pkg/informer/informer.go

```diff
@@ -126,108 +158,150 @@ func NewDynamicDiscoverySharedInformerFactory(
+	}
+
+	return NewGenericDiscoveringDynamicSharedInformerFactory[cache.SharedIndexInformer, cache.GenericLister](
+		func(gvr schema.GroupVersionResource, resyncPeriod time.Duration, indexers cache.Indexers) informers.GenericInformer {
+			return dynamicinformer.NewFilteredDynamicInformer(
+				dynamicClient,
+				gvr,
+				"",
+				resyncPeriod,
+				indexers,
+				tweakListOptions,
+			)
+		},
+		filterFunc,
+		gvrSource,
+		indexers,
+	)
 }
 
-func byGroupVersionResourceIndexFunc(obj interface{}) ([]string, error) {
-	crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
-	if !ok {
-		return nil, fmt.Errorf("%T is not a CustomResourceDefinition", obj)
+// NewSelfDiscoveringDynamicSharedInformerFactory returns a factory for cluster-aware
+// shared informers that discovers new types and informs on updates to resources of
+// those types.
+// It retrieves GVR-related information from CRD informers.
+func NewSelfDiscoveringDynamicSharedInformerFactory(
+	cfg *rest.Config,
+	filterFunc func(obj interface{}) bool,
+	crdInformer kcpapiextensionsv1informers.CustomResourceDefinitionClusterInformer,
+	indexers cache.Indexers,
+) (*DiscoveringDynamicSharedInformerFactory, error) {
+	cfg = rest.AddUserAgent(rest.CopyConfig(cfg), "kcp-partial-metadata-informers")
+	metadataClusterClient, err := metadataclient.NewDynamicMetadataClusterClientForConfig(cfg)
+	if err != nil {
+		return nil, err
 	}
 
-	var ret []string
-
-	for _, v := range crd.Spec.Versions {
-		if !v.Served {
-			continue
-		}
-		ret = append(ret, byGroupVersionResourceKeyFunc(crd.Spec.Group, v.Name, crd.Spec.Names.Plural))
+	crdGVRSource, err := newCRDGVRSource(crdInformer.Informer())
+	if err != nil {
+		return nil, err
 	}
 
-	return ret, nil
+	return NewDiscoveringDynamicSharedInformerFactory(metadataClusterClient, filterFunc, nil, crdGVRSource, indexers)
+}
+
+// NewDiscoveringDynamicSharedInformerFactory returns a factory for cluster-aware
+// shared informers that discovers new types and informs on updates to resources of
+// those types.
+// It receives the GVR-related information by delegating to a GVRSource.
+func NewDiscoveringDynamicSharedInformerFactory(
+	dynamicClusterClient kcpdynamic.ClusterInterface,
+	filterFunc func(obj interface{}) bool,
+	tweakListOptions dynamicinformer.TweakListOptionsFunc,
+	gvrSource GVRSource,
+	indexers cache.Indexers,
+) (*DiscoveringDynamicSharedInformerFactory, error) {
+	f, err := NewGenericDiscoveringDynamicSharedInformerFactory[kcpcache.ScopeableSharedIndexInformer, kcpcache.GenericClusterLister](
+		func(gvr schema.GroupVersionResource, resyncPeriod time.Duration, indexers cache.Indexers) kcpinformers.GenericClusterInformer {
+			indexers[kcpcache.ClusterIndexName] = kcpcache.ClusterIndexFunc
+			indexers[kcpcache.ClusterAndNamespaceIndexName] = kcpcache.ClusterAndNamespaceIndexFunc
+			return kcpdynamicinformer.NewFilteredDynamicInformer(
+				dynamicClusterClient,
+				gvr,
+				resyncPeriod,
+				indexers,
+				nil,
```

Comment:
What is this `nil`? Why?

Reply:
Should be `tweakListOptions`. I'll fix it.

Reply:
Fixed